### PR TITLE
Enable nullability in DesignerHost and SiteNestedContainer

### PR DESCRIPTION
**Pull request by halgab**
## Proposed changes

- Enable nullability in `DesignerHost` and `SiteNestedContainer` 
- Change `Hashtable` with `Dictionary<object, object>` in `DesignerHost` (contributes to #8143)
- Stop throwing null-reference exceptions when adding or removing null components from `DesignHost` and `SiteNestedContainer`. The idea behind this change (that is obviously open for debate) is that it makes those method match the behaviour of the ones they override (from class `Container`). This one is the second commit, that can easily be rollbacked.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9340)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs [hunk 1]
```diff
@@ -17,18 +14,18 @@ internal sealed partial class DesignerHost
     internal class Site : ISite, IServiceContainer, IDictionaryService
     {
         private readonly IComponent _component;
-        private Hashtable _dictionary;
+        private Dictionary<object, object>? _dictionary;
```
Comment (elachlan) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs:17]: Are the usages of the dictionary not strongly typed?

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs:17]: It supports the implementation of the weakly-typed `IDictionaryService` interface. Removing the `Hashtable` in that case is less interesting, but at least it lets us specify that the dictionary values are not null.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.DesignerHostTransaction.cs [hunk 2]
```diff
@@ -13,14 +11,15 @@ internal sealed partial class DesignerHost
...
     {
-        private DesignerHost _host;
+        private DesignerHost? _host;
+        private readonly Stack<DesignerTransaction> _transactions;
 
         public DesignerHostTransaction(DesignerHost host, string description) : base(description)
         {
             _host = host;
-            _host._transactions ??= new Stack<DesignerTransaction>();
+            _transactions = _host._transactions ??= new Stack<DesignerTransaction>();
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.DesignerHostTransaction.cs:20]: isn't `_host._transactions` going out sync here by operating on local? what if caller query on transactions?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs [hunk 3]
```diff
@@ -39,9 +36,9 @@ private IServiceContainer SiteServiceContainer
         {
             get
             {
-                SiteNestedContainer nc = ((IServiceProvider)this).GetService(typeof(INestedContainer)) as SiteNestedContainer;
+                SiteNestedContainer nc = (((IServiceProvider)this).GetService(typeof(INestedContainer)) as SiteNestedContainer)!;
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs:39]:
```suggestion
                SiteNestedContainer? nc = (((IServiceProvider)this).GetService(typeof(INestedContainer)) as SiteNestedContainer);
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs [hunk 4]
```diff
@@ -39,9 +36,9 @@ private IServiceContainer SiteServiceContainer
...
                 Debug.Assert(nc is not null, "We failed to resolve a nested container.");
-                IServiceContainer sc = nc.GetServiceInternal(typeof(IServiceContainer)) as IServiceContainer;
+                IServiceContainer sc = (nc.GetServiceInternal(typeof(IServiceContainer)) as IServiceContainer)!;
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs:41]:
```suggestion
                IServiceContainer? sc = (nc.GetServiceInternal(typeof(IServiceContainer)) as IServiceContainer);
```

Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs:41]: Debug assert should help compiler suppress any further warnings with nullability here.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs [hunk 5]
```diff
@@ -70,22 +67,22 @@ object IDictionaryService.GetKey(object value)
         /// <summary>
         ///  Retrieves the value corresponding to the given key.
         /// </summary>
-        object IDictionaryService.GetValue(object key)
+        object? IDictionaryService.GetValue(object key)
         {
-            if (_dictionary is not null)
+            if (_dictionary is null || !_dictionary.TryGetValue(key, out object? value))
             {
-                return _dictionary[key];
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs:77]: Isn't this throwing key not found exception earlier and now return `null`?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs [hunk 6]
```diff
@@ -1219,11 +1183,10 @@ void IDesignerLoaderHost.EndLoad(string rootClassName, bool successful, ICollect
                     _state[s_stateLoading] = true;
                     Unload();
 
-                    errorCollection = errorCollection is null ? new List<object>() : errorCollection.Cast<object>().ToList();
-                    if (errorCollection is List<object> errorList)
-                    {
-                        errorList.Insert(0, ex);
-                    }
+                    List<object> errorList = errorCollection is null ? new List<object>() : errorCollection.Cast<object>().ToList();
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs:1186]:
```suggestion
                    List<object> errorList = errorCollection is null ? new() : errorCollection.Cast<object>().ToList();
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs (hunk 4 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs:41]: Okay thanks. I had some doubts about it in Release. Wiil fix

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.DesignerHostTransaction.cs (hunk 2 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.DesignerHostTransaction.cs:20]: `_host._transactions` is updated only here, so it only should have one instance and this should be safe. I'll rollback anyway to make it clear

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs (hunk 5 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.Site.cs:77]: `_dictionary` was a `Hashtable`, so it used to return `null` when the key wasn't present

**Review by dreddy-work — APPROVED** [review]

**Pull request merged**